roles/kafka_controller/tasks/main.yml

```diff
@@ -0,0 +1,435 @@
+    group: "{{kafka_controller_group}}"
+  when: kafka_controller_jolokia_enabled|bool
+  notify: restart kafka
+  diff: "{{ not mask_sensitive_diff|bool }}"
+  tags:
+    - configuration
+
+- name: Create Kafka Controller Jaas Config
+  template:
+    src: kafka_server_jaas.conf.j2
+    dest: "{{kafka_controller.jaas_file}}"
+    mode: 0640
+    owner: "{{kafka_controller_user}}"
+    group: "{{kafka_controller_group}}"
+  when:
+    - "'GSSAPI' in kafka_controller_sasl_enabled_mechanisms"
+  notify: restart kafka
+  tags:
+    - configuration
+
+- name: Deploy JMX Exporter Config File
+  template:
+    src: "{{kafka_controller_jmxexporter_config_source_path}}"
+    dest: "{{kafka_controller_jmxexporter_config_path}}"
+    mode: 0640
+    owner: "{{kafka_controller_user}}"
+    group: "{{kafka_controller_group}}"
+  when: kafka_controller_jmxexporter_enabled|bool
+  tags:
+    - configuration
+
+- name: Create Service Override Directory
+  file:
+    path: "{{kafka_controller.systemd_override | dirname }}"
+    owner: "{{kafka_controller_user}}"
+    group: "{{kafka_controller_group}}"
+    state: directory
+    mode: 0640
+  tags:
+    - systemd
+
+- name: Write Service Overrides
+  template:
+    src: override.conf.j2
+    dest: "{{ kafka_controller.systemd_override }}"
+    mode: 0640
+    owner: root
+    group: root
+  notify: restart kafka
+  diff: "{{ not mask_sensitive_diff|bool }}"
+  tags:
+    - systemd
+
+- name: Create sysctl directory on Debian distributions
+  file:
+    path: /usr/lib/sysctl.d
+    state: directory
+    mode: 0755
+  when: ansible_distribution == "Debian"
+  tags:
+    - sysctl
+    - privileged
+
+- name: Tune virtual memory settings
+  sysctl:
+    name: "{{ item.key }}"
+    value: "{{ item.value }}"
+    state: present
+    sysctl_file: "{{ kafka_controller_sysctl_file }}"
+    reload: true
+  with_dict: "{{ kafka_controller_sysctl }}"
+  tags:
+    - sysctl
+    - privileged
+
+- name: Certs were Updated - Trigger Restart
+  command: /bin/true
+  notify: restart kafka
+  when: certs_updated|bool
+
+# We need to start MDS here to proceed with secrets protection
+- meta: flush_handlers
+
+# Starting confluent cli v2, secrets protection works only when rbac
+- name: Encrypt secrets
+  include_tasks: secrets_protection.yml
+  when:
+    - kafka_controller_secrets_protection_enabled|bool or kafka_controller_client_secrets_protection_enabled|bool
+    - rbac_enabled|bool
+
+- meta: flush_handlers
+
+- name: Kafka Started
+  systemd:
+    name: "{{kafka_controller_service_name}}"
+    enabled: true
+    state: started
+  tags:
+    - systemd
+
+- name: Wait for Controller health checks to complete
+  include_tasks: health_check.yml
+  when:
+    - kafka_controller_health_checks_enabled|bool
+    - not ansible_check_mode
+  tags: health_check
+
+- name: Register Cluster
+  include_tasks: register_cluster.yml
+  when: kafka_controller_cluster_name|length > 0 and rbac_enabled|bool
+
+- name: Create RBAC Rolebindings
```

Comment:
How do these tasks run if the MDS has not started yet?

Reply:
yes you're right, we're currently working on rbac in kraft. The code in this PR does not support rbac, will raise another PR for rbac support. Also, these tasks would run when we have external_mds_enabled since in that case MDS would be up and running on another cluster.